src/xml_parser_main_metadata_together.c

```diff
@@ -0,0 +1,371 @@
+        // package was already encountred in some other metadata type
+
+        if (pval == NULL) {
+            assert(0);
+            //this pkg was already fully handled
+            //TODO(amatej): maybe error?
+        } else {
+            *pkg = (cr_Package *) pval;
+        }
+    } else {
+        // we are handling never before seen package
+
+        if (cb_data->newpkgcb) {
+            // call user specified new function if provided
+            if (cb_data->newpkgcb(pkg, pkgId, name, arch, cb_data->newpkgcb_data, err)) {
+                return CR_CB_RET_ERR;
+            }
+        } else {
+            *pkg = cr_package_new();
+        }
+        g_hash_table_insert(cb_data->in_progress_pkgs, g_strdup(pkgId), *pkg);
+    }
+
+    return CR_CB_RET_OK;
+}
+
+static int
+filelists_pkgcb(cr_Package *pkg, void *cbdata, G_GNUC_UNUSED GError **err)
+{
+    cr_CbData *cb_data = cbdata;
+    cb_data->filelists_count++;
+    pkg->loadingflags |= CR_PACKAGE_LOADED_FIL;
+    return call_user_callback_if_package_finished(pkg, cb_data, err);
+}
+
+static int
+other_pkgcb(cr_Package *pkg, void *cbdata, G_GNUC_UNUSED GError **err)
+{
+    cr_CbData *cb_data = cbdata;
+    cb_data->other_count++;
+    pkg->loadingflags |= CR_PACKAGE_LOADED_OTH;
+    return call_user_callback_if_package_finished(pkg, cb_data, err);
+}
+
+static int
+primary_pkgcb(cr_Package *pkg, void *cbdata, G_GNUC_UNUSED GError **err)
+{
+    cr_CbData *cb_data = cbdata;
+    cb_data->primary_count++;
+
+    gpointer pkey = NULL;
+    gpointer pval = NULL;
+    if (g_hash_table_lookup_extended(cb_data->in_progress_pkgs, pkg->pkgId, &pkey, &pval)) {
+        // package was already encountred in some other metadata type
+
+        if (pval == NULL) {
+            //this pkg was already fully handled
+            //TODO(amatej): maybe error?
+            assert(0);
+        } else {
+            cr_package_copy_into(pkg, pval);
+            cr_package_free(pkg);
+            pkg = pval;
+        }
+    } else {
+        // we are handling never before seen package
+
+        if (cb_data->newpkgcb) {
+            // call user specified new function if provided and copy q
+            cr_Package *user_created_pkg;
+            if (cb_data->newpkgcb(&user_created_pkg, pkg->pkgId, NULL, NULL, cb_data->newpkgcb_data, err)) {
+                return CR_CB_RET_ERR;
+            } else {
+                cr_package_copy_into(pkg, user_created_pkg);
+                cr_package_free(pkg);
+                pkg = user_created_pkg;
+            }
+        }
+        g_hash_table_insert(cb_data->in_progress_pkgs, g_strdup(pkg->pkgId), pkg);
+    }
+
+    pkg->loadingflags |= CR_PACKAGE_FROM_XML;
+    pkg->loadingflags |= CR_PACKAGE_LOADED_PRI;
+
+    return call_user_callback_if_package_finished(pkg, cb_data, err);
+}
+
+static int
+parse_next_section(CR_FILE *target_file, const char *path, cr_ParserData *pd, GError **err)
+{
+    char buf[XML_BUFFER_SIZE];
```

Comment:
So, instead of parsing from start to end element of a package one package at a time, we parse the file one chunk of bytes at a time, which could be >1 package or <1 package, and just keep track of which packages are "done" -- yes?

Reply:
This makes a lot of sense given the design constraints you're dealing with (libxml, existing parsers written in SAX style).  I wasn't sure whether it could be done without a "pull" / "StAX" parser, which I don't think libxml provides, so I thought it might require a more fundamental change to the architecture.  

There's some limitations to the approach, but it will work, and it's nice to see that it didn't end up requiring a complete rewrite.

Reply:
You are correct. To be honest I didn't know that "pull" / "StAX" parsing models were a thing. Now I managed to find this libxml2 API: http://xmlsoft.org/xmlreader.html, I had no idea about this. :slightly_frowning_face: 
It would be a big rewrite but it looks like its the better solution and it seems DTD validation is also easier with that approach.

I will look into it some more, but I am not sure if I will be able to do it right now.

Reply:
No worries, I agree that it would basically represent a complete rewrite.  This is probably the better  solution in the short term and there's now that we have it.

I was actually experimenting with this method in a side project written in Rust, but I haven't really had time to work on it in months, so it's still pretty limited (no streaming one package at a time, although it would definitely be possible eventually).  I think it would mostly translate to C (if you eventually wanted to use it as a reference).  

https://github.com/dralley/rpmrepo_rs/blob/master/rpmrepo_metadata/src/primary.rs#L153-L198
https://github.com/dralley/rpmrepo_rs/blob/master/rpmrepo_metadata/src/primary.rs#L400-L604

Reply:
FYI I implemented the Rust version of package streaming one-by-one in a couple of hours.